autogpt_platform/backend/backend/blocks/smart_decision_maker.py

```diff
@@ -0,0 +1,184 @@
+from backend.data.graph import Graph
+from backend.data.model import SchemaField
+
+logger = logging.getLogger(__name__)
+
+
+@thread_cached
+def get_database_manager_client():
+    from backend.executor import DatabaseManager
+    from backend.util.service import get_service_client
+
+    return get_service_client(DatabaseManager)
+
+
+class SmartDecisionMakerBlock(Block):
+    # Note: Currently proving out the concept of determining the inputs a tool takes
+
+    class Input(BlockSchema):
+        # Note: This is a placeholder for the actual input schema
+        text: str = SchemaField(description="The text to print to the console.")
+
+    class Output(BlockSchema):
+        # Starting with a single tool.
+        tool: dict[str, dict[str, Any]] = SchemaField(
+            description="The tools that are available to use."
+        )
+
+    def __init__(self):
+        super().__init__(
+            id="3b191d9f-356f-482d-8238-ba04b6d18381",
+            description="Uses AI to intelligently decide what tool to use.",
+            categories={BlockCategory.BASIC},
+            input_schema=SmartDecisionMakerBlock.Input,
+            output_schema=SmartDecisionMakerBlock.Output,
+            test_input={"text": "Hello, World!"},
+            test_output=("status", "printed"),
+        )
+
+    @staticmethod
+    def _create_function_signature(node_id: str, graph: Graph) -> list[dict[str, Any]]:
+        """
+        Creates a list of function signatures for tools linked to a specific node in a graph.
+
+        This method identifies all tool links associated with a given node ID within a graph,
+        groups them by tool name, and constructs a function signature for each tool. Each
+        function signature includes the tool's name, description, and parameters required
+        for its execution.
+
+        Parameters:
+        - node_id (str): The ID of the node for which tool function signatures are to be created.
+        - graph (GraphModel): The graph model containing nodes and links.
+
+        Returns:
+        - list[dict[str, Any]]: A list of dictionaries, each representing a tool function signature.
+          Each dictionary contains:
+            - "name": The name of the tool.
+            - "description": A description of the tool or its multi-step process.
+            - "parameters": A dictionary detailing the parameters required by the tool, including:
+                - "type": The data type of the parameter.
+                - "description": A description of the parameter.
+
+        Raises:
+        - ValueError: If no tool links are found for the given node ID or if no tool sink nodes are identified.
+        """
+        # Filter the graph links to find those that are tools and are linked to the specified node_id
+        tool_links = [
+            link
+            for link in graph.links
+            if link.source_name == "tool" and link.source_id == node_id
```

Comment:
instead of pulling the whole graph you can use `get_node` function and find the output_links of the "tool" source.

Reply:
I see, if you need some detail I think you need to make the function for it, sth like:

```
async def get_connected_output_nodes(source_node_id: str) -> list[Node]:
    links = await AgentNodeLink.prisma().find_many(
        where={"agentNodeSourceId": source_node_id},
        include={"AgentNode": {"include": AGENT_NODE_INCLUDE}},
    )
    return [NodeModel.from_db(link.AgentNodeSink) for link in links]
```

You can still do one query, and avoid pulling the whole graph in case it's a large graph.

Reply:
Anyway the idea looks good though :+1: